First commits

michaelsaki opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> Breakout cyhy-nvdsync script into AWS Lambda #

## 🗣 Description ##

 The cyhy-nvdsync script in [cisagov/cyhy-core](https://github.com/cisagov/cyhy-core) has been converted into an AWS Lambda. The lambda will retrieve the data from a remote source. It will process the data into CVE Documents to be stored in a MongoDB, like the original cyhy-nvdsync script. It has been improved to run in Python3 and also runs asynchronously with the new Beanie ODM.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

Given what the original cyhy-nvdsync script did:

- Retrieve data from a remote source.
- Minimally process the data.
- Add documents with this processed data to the appropriate collection.

It made sense to convert it into a Lambda. This aligns with other Lambdas in the CyHy system like [cisagov/assessment-data-import ](https://github.com/cisagov/assessment-data-import)and [cisagov/findings-data-import](https://github.com/cisagov/findings-data-import-lambda). This removes the need to maintain an EC2 instance to run these tasks.

This also resolves the issue of converting the cyhy-nvdsync script from Python2 to Python3 as referenced in https://github.com/cisagov/cyhy-core/issues/73

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

For testing I locally stood up a MongoDB container and invoked the Lambda using Cloudwatch. The Lambda runs for about 10-15 minutes as it processes each CVE Document (over 200,000) into the database.

I also made local unit tests to ensure that the Lambda will properly remove any "unwanted" or "outdated" CVE Documents that might already exist in the database. I did this by disabling the filters for the CVE Documents so all the data from the NVD remote site is inserted into the database. Then I would build a new image for the Lambda with the filters enabled, ran it with the database that I loaded all the Documents onto, and checked to make sure the Lambda would remove the Documents from the database.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release.

Commit:
Remove skeleton Lambda implementation

Pare the lambda_handler.py file down to the handler function and
the logging boilerplate to begin development.

Commit:
Initial implementation

Implement the initial functionality to read the CISA NVD JSON, save the
CVE IDs it contains to a MongoDB collection. Since we are using the Beanie
library to provide a MongoDB ODM we use Python's asyncio features for
the implementation. NOTE: This initial implementation isn't parsing the
URLs correctly. Will rebase once functionality is correct.

Commit:
Add more fields to CVEDoc and process JSON URLs

The current version of this code doesn't function like what is expected.
The processing of JSON URLs isn't correctly parsing the CVEs. Will refactor
once everything is functioning like it should but committing for now to
have work visible to Fusion team.

Commit:
refactor the parse_cve_json()

Attempted to refactor the parse_cve_json function to
replicate initial cyhy-nvdsync. This is causing issues
with Beanie ODM because the original script didn't have
async functionality which is inherent to Beanie.

Commit:
Make script compatible with beanie

Previous commmit had code that wasn't compatible with
ODM Beanie. Made changes to follow proper Beanie formatting.
Script is still not processing JSON URLs properly.

Commit:
Revert commented code and fields for CVE document

Commit:
Fix fields to the CVE Documents

Fields for the CVE Documents weren't being populated correctly.
Made adjustments to these fields and how they were being stored
in the MongoDB. This code still has issues when it comes to
processing the CVE Documents asynchonously. Logs from
local tests make it seem like a few of the CVE Docs are
succesfully processed only to have the remaining ones
fail.

Commit:
Fix more of the CVE fields

Initially these fields were not
optional which led to issues when
trying to save them into the database.
The CVE Documents won't always have
every field filled out from the
JSON files.

Commit:
Change the filtering of CVE Documents

The CVE Documents are not being filtered properly
it appears that only CVE scores from 1999 to 2002
are being successfully stored in the database until
the script eventually times out and crashes.

Commit:
Add breadcrumbs for debugging async collision

The lambda is still timing out. I suspect that the
Issue is occurring from collision between async
functions but haven't confirmed this assumption.
Breadcrumbs have been added to the lambda to
attempt to debug the issue through console logs.

Commit:
Fix delete function

The lambda will now properly remove outdated
CVE Documents from the DB if they already exist.

Commit:
Add new event loop before lambda is executed

Event loop would stay open and cause the
lambda to crash. Now a new event loop
will open in between each invocation
of the lambda.

Commit:
Add time for lambda, retry for URL, and ssm params

The nvdsync lambda is now functional but the ssm
parameters haven't officially been tested. The implementation
was copied from the kevsync lambda so it should be
good to go. Also added retries for the URL socket
connection. The default is set for 10 attempts with
5 second sleeps in between each attempt. Also the
lambda will timeout with the default 5 minute window.
Increased the time so the lambda can run until all
the CVEs are processed into the database.

jsf9k committed:
Make a minor formatting change to satisfy the Black linter

jsf9k commented:
@michaelsaki - I merged the Lineage changes into this repo and rebased this branch on top of them.  I also added a very minor commit to make the GH Actions linters pass.

jsf9k commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,208 @@
+        resp = ssm_client.get_parameter(Name=parameter_name, WithDecryption=True)
+        return resp["Parameter"]["Value"]
+    except ClientError:
+        logging.error("Unable to retrieve SSM Parameter Store key: %s", parameter_name)
```
I think we can dismiss this alert since it's only logging the parameter name, but I'd like to know if anyone else in @cisagov/team-ois sees a concern here.

dav3r replied:
Yup, I'm in agreement with you @jsf9k.

jsf9k commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,207 @@
+        """If the CVE is in the database, it needs to be deleted from the database"""
 
+        if await CVEDoc.find_one(CVEDoc.id == cve_id):
+            print("x", end="")
```
Are these `print()` lines intended to be left in?

michaelsaki replied:
@jsf9k I asked the same question to @mcdonnnj. He said "That will end up in CloudTrail and will be there for confirmation/debugging still."

michaelsaki replied:
@jsf9k I had previously marked this as "resolved" so I recently unmarked it so you could officially mark it.

jsf9k replied:
@mcdonnnj - Can you comment here?  Are the `print()` lines intended to be left in?

dav3r replied:
Pinging @mcdonnnj again here - we just want to get your confirmation on this thread.  I believe it makes sense to keep these `print()` lines in here for auditing purposes.

dav3r replied:
@jsf9k I think it's safe to resolve this conversation- 7 weeks is too long to wait for an answer to this.

jsf9k commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,207 @@
-        result["message"] = cowsay.get_output_string(character, contents)
 
-    return result
+        if not cve_id:
+            raise ValueError("JSON does not look like valid CISA CVE data.")
```
Should this be earlier in the function, say just after you set `cve_id` in line 75?

michaelsaki replied:
This has been addressed in the latest push. https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/1220b72e11ee456746d6f29feab755e09d089c8d

jsf9k commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,207 @@
-
-def handler(event, context) -> dict[str, Optional[str]]:
+async def process_cve_json(json_stream) -> None:
+    """Process the provided CVEs JSONs and update the database with its contents."""
```
```suggestion
    """Process the provided CVEs JSONs and update the database with their contents."""
```

michaelsaki replied:
This has been addressed in the latest push. https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/1220b72e11ee456746d6f29feab755e09d089c8d

jsf9k commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,207 @@
+    if data.get("CVE_data_type") != "CVE":
+        raise ValueError("JSON does not look like valid NVD CVE data.")
+
+    tasks = [asyncio.create_task(process_nvd(cve)) for cve in data.get("CVE_Items", [])]
+
+    for task in asyncio.as_completed(tasks):
+        await task
```
Would it be simpler to use a `TaskGroup` here?

jsf9k replied:
`async.gather()` is the old way, from before `TaskGroup` became a thing.  You can just use that.

mcdonnnj replied:
Task groups were I believe introduced in [Python 3.11](https://www.python.org/downloads/release/python-3110/) no? Until the way we do Lambdas is updated (work in progress) the repos are still designed to support multiple Python versions.

michaelsaki replied:
This has been fixed in the latest push. https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/1220b72e11ee456746d6f29feab755e09d089c8d

jsf9k commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,207 @@
+
+    for cve_url in cve_urls:
+        # We disable the bandit blacklist for the urllib.request.urlopen() function
+        # because the URL is either the defaul (safe) URL or one provided in the
```
```suggestion
        # because the URL is either the default (safe) URL or one provided in the
```

michaelsaki replied:
This has been addressed in the latest push. https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/1220b72e11ee456746d6f29feab755e09d089c8d

jsf9k commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,207 @@
+
+        buf = BytesIO(socket.read())
+        f = gzip.GzipFile(fileobj=buf)
+        await process_cve_json(f)
```
You went through the trouble to use `async` throughout this PR, but it seems like your use of `await` is almost always such that it forces synchronous behavior.  Did you want to change the logic to schedule _all_ the `process_cve_json(f)` tasks and then wait for them _all_ to finish?  That would allow more than one to run concurrently.

michaelsaki replied:
@mcdonnnj  I think you had said that we want each CVE year to run till it fully completes before running the next year in the process_cve_json(f) correct?

dav3r replied:
What is the reason why we wouldn't want to pull in multiple years of CVEs concurrently?

michaelsaki replied:
@dav3r and @jsf9k I had to make a few fairly big changes to how we were reading from the URLs to get the CVEs to run concurrently. Before we were `url.request.urlopen(cve_url)` but when trying to group the reading tasks together I kept running into an error `IncompleteRead(xxxxxx bytes read, xxxxxx more expected)`. I switched to using `aiohttp` to read from the URLs you will see it here [Lines 147 - 150](https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b#diff-4655711b74e9f79a2566fdbac7451f4dcc265a12d3ad743e519353283898a155R147-R150) and also here [Lines 157 - 188](https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b#diff-4655711b74e9f79a2566fdbac7451f4dcc265a12d3ad743e519353283898a155R157-R188). This is all under the commit https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b

jsf9k commented on src/lambda_handler.py:
```diff
@@ -109,19 +229,75 @@ def handler(event, context) -> dict[str, Optional[str]]:
+    # but we verify that there are only str elements before this point.
+    mongodb_uri = build_mongodb_uri(*[v for k, v in mongodb_uri_elements])  # type: ignore
+
+    if mongodb_uri is None:
+        logging.error("Unable to import KEV data.")
+        return
+
+    # Set up the Motor session if necessary
+    if motor_client is None:
+        motor_client = AsyncIOMotorClient(mongodb_uri)
+
+    try:
+        asyncio.get_event_loop().run_until_complete(
+            process_urls(cve_json_urls, write_db)
+        )
```
Would it be simpler to use a `TaskGroup` here?

jsf9k replied:
`async.gather()` is the old way, from before `TaskGroup` became a thing.  You can just use that.

mcdonnnj replied:
Task groups were I believe introduced in [Python 3.11](https://www.python.org/downloads/release/python-3110/) no? Until the way we do Lambdas is updated (work in progress) the repos are still designed to support multiple Python versions.

michaelsaki replied:
I am not sure if we can put a `TaskGroup` or rather `async.gather()`. We would need to make the lambda `handler` function async and when I attempted to do that I kept running into errors. I imagine it is possible to get it to run asynchronously but it wasn't a simple change without looking under the hood of how we are running the lambda handler.

jsf9k replied:
Can you just add a comment that this is icky and we should revisit it to find a better solution?

michaelsaki replied:
@jsf9k I addressed this here: [e6397c6](https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b#diff-4655711b74e9f79a2566fdbac7451f4dcc265a12d3ad743e519353283898a155L302-R321)

jsf9k reviewed (commented):
I left a few comments for discussion.

michaelsaki commented:
In regards to the comments made about using TaskGroups:
Currently the cyhy-nvdsync-lambda isn't written in Python 3.11 (when TaskGroups were introduced). The latest version of Python that the lambda runs under is Python 3.9. We can adjust when the lambda skeleton is updated to 3.11 but for now I will try to implement a similar functionality to TaskGroups.

jsf9k commented:
> In regards to the comments made about using TaskGroups: Currently the cyhy-nvdsync-lambda isn't written in Python 3.11 (when TaskGroups were introduced). The latest version of Python that the lambda runs under is Python 3.9. We can adjust when the lambda skeleton is updated to 3.11 but for now I will try to implement a similar functionality to TaskGroups.

`async.gather()` is the old way, from before `TaskGroup` became a thing.  You can just use that.

Commit:
Add task groups to optimize async

Note that we aren't actually using
Task Groups as they were introduced
in Python 3.10 and this is currently
written in 3.9. We use asyncio.gather
instead. But now we can run tasks
asynchronously.

michaelsaki commented:
@mcdonnnj  and @jsf9k My latest push (https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/1220b72e11ee456746d6f29feab755e09d089c8d) should address all the concerns that were mentioned in the discussions above. I believe we can move towards closing this PR.

dav3r commented on docker-compose.yml:
```diff
@@ -19,6 +19,8 @@ services:
-    image: cisagov/example_lambda:${LAMBDA_TAG:-latest}
+    image: cisagov/nvdsync_lambda:${LAMBDA_TAG:-latest}
+    environment:
+      - AWS_LAMBDA_FUNCTION_TIMEOUT=900000
```
It never hurts to add a comment like this:
```suggestion
      # 900,000 milliseconds = 15 minutes
      - AWS_LAMBDA_FUNCTION_TIMEOUT=900000
```

michaelsaki replied:
@dav3r I addressed this here: https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/5764f7c19a4cb425074d410a44b1f8ec36ef858d

jsf9k replied:
@michaelsaki - Why are you averse to simply committing the suggestion via @dav3r's comment?

michaelsaki replied:
> @michaelsaki - Why are you averse to simply committing the suggestion via @dav3r's comment?

@jsf9k - I wasn't aware that you could simply commit them through the suggestion. Thank you for pointing out the capability in GitHub. Will definitely be using it in the future!

dav3r commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,215 @@
-    return result
+async def process_nvd(cve) -> str:
+    """Add the provided CVE to the database and return its id."""
+    # Fill fields for the CVE Documents from the given JSON files
```
Remove unnecessary capitalization:
```suggestion
    # Fill fields for the CVE documents from the given JSON files
```

michaelsaki replied:
@dav3r I capitalized this because it was the name of the Beanie ODM class Document. See here https://beanie-odm.dev/tutorial/defining-a-document/ 

That being said they do call the section "Defining a _document_" with no capitalization so I am fine either way but still thought it was worth mentioning.

dav3r replied:
Understood, but in this comment, I still think `documents` reads more cleanly than `Documents`.  Go with whatever your ❤️  tells you and I will support it.

jsf9k replied:
https://www.youtube.com/watch?v=yCC_b5WHLX0

dav3r replied:
https://www.youtube.com/watch?v=Ud6sU3AclT4

michaelsaki replied:
@jsf9k and @dav3r I feel Rick Rolled having clicked on both of those links haha

michaelsaki replied:
@dav3r I addressed this here [Line 75](https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b#diff-4655711b74e9f79a2566fdbac7451f4dcc265a12d3ad743e519353283898a155R75) I went with my ❤️

dav3r replied:
Thanks for taking care of this @michaelsaki.  I do have to point out that https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b is a big, gnarly commit that does a bunch of different things.  In the future, please try to keep your commits smaller and more focused on a single thing.  It makes them easier to review and understand.

michaelsaki replied:
@dav3r You are right I did think that after I had made the commit. I will do better next time.

dav3r commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,215 @@
+async def process_nvd(cve) -> str:
+    """Add the provided CVE to the database and return its id."""
+    # Fill fields for the CVE Documents from the given JSON files
+    cve_id = cve["cve"]["CVE_data_meta"]["ID"]
+    if not cve_id:
+        raise ValueError("JSON does not look like valid CISA CVE data.")
```
Shouldn't this be in a `try/except` block?  If the JSON is malformed (e.g. missing any of the "cve", "CVE_data_meta", or "ID" fields), we will get a `KeyError` and we'll want to log what the `cve` object looks like.  Another possibility is if the fields are all there, but the "ID" field is empty.

This check and corresponding error message(s) should be improved here and tested to ensure they work correctly.

michaelsaki replied:
@dav3r I addressed this here [Lines 76-85](https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b#diff-4655711b74e9f79a2566fdbac7451f4dcc265a12d3ad743e519353283898a155R76-R85)

dav3r replied:
Looks much better now. 👍

dav3r commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,215 @@
-    return result
+        # Gather all saving tasks to run concurrently
+        tasks = [asyncio.create_task(cve_doc.save())]
+        await asyncio.gather(*tasks)
```
Why do we want to wait and do all of our DB saving concurrently?

michaelsaki replied:
@dav3r at first I thought it was processing the documents faster but you were right to question it. Further testing showed that it only slowed things down. I addressed it here [Line 125](https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b#diff-4655711b74e9f79a2566fdbac7451f4dcc265a12d3ad743e519353283898a155R125)

dav3r replied:
Excellent! 🚀

dav3r commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,215 @@
+
+
+async def process_urls(cve_urls, db) -> None:
+    """Initialize beanie ODM and begin processing CVE URLs."""
```
Use proper capitalization of package name:
```suggestion
    """Initialize Beanie ODM and begin processing CVE URLs."""
```

michaelsaki replied:
@dav3r I addressed this here [Line 154](https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b#diff-4655711b74e9f79a2566fdbac7451f4dcc265a12d3ad743e519353283898a155R154)

dav3r reviewed (changes requested):
I noted a few small items that should be addressed.

Commit:
Update dependency to include aiohttp

Commit:
Add a comment for lambda timeout

Commit:
Add fixes from PR 'First commits' feedback

Changes include the following:
1. Grammar has been improved on comments.
2. CVE years are now running concurrently.
   Had to switch to aiohttp to read from
   the URLs.
3. Additional error handling has been
   added for CVE processing.
4. Added more comments for issues that
   will need to be addressed in the future.

jsf9k commented on src/py3.8/Pipfile:
```diff
@@ -7,5 +7,10 @@ name = "pypi"
+beanie = "*"
+boto3 = "*"
+botocore = "*"
+aiohttp = "*"
```
Alphabetization:
```suggestion
aiohttp = "*"
beanie = "*"
boto3 = "*"
botocore = "*"
```

michaelsaki replied:
@jsf9k Addressed changes here https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/cc057b25eec7f2b1d5e8af5c282867ef5bab8d0f

jsf9k commented on src/py3.9/Pipfile:
```diff
@@ -7,5 +7,10 @@ name = "pypi"
+beanie = "*"
+boto3 = "*"
+botocore = "*"
+aiohttp = "*"
```
Alphabetization:
```suggestion
aiohttp = "*"
beanie = "*"
boto3 = "*"
botocore = "*"
```

michaelsaki replied:
@jsf9k Addressed changes here https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/17146fa1d7a70acf64517e100e51cc5f463a771e

dav3r commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,231 @@
-        failed_task(result, error_msg)
     else:
+        version = "V3" if "baseMetricV3" in cve["impact"] else "V2"
+        cvss_base_score = cve["impact"]["baseMetric" + version]["cvss" + version][
+            "baseScore"
+        ]
+        cvss_version_temp = cve["impact"]["baseMetric" + version]["cvss" + version][
+            "version"
+        ]
+
+        # Fill document fields with CVE data
+        print(".", end="")
+
         try:
```
This `try:` block needs to begin earlier if we want to catch malformed `cve` JSON data:
```suggestion
        try:
            cvss_base_score = cve["impact"]["baseMetric" + version]["cvss" + version][
                "baseScore"
            ]
            cvss_version_temp = cve["impact"]["baseMetric" + version]["cvss" + version][
                "version"
            ]

            # Fill document fields with CVE data
            print(".", end="")
```

dav3r commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,231 @@
-    return result
+            while True:
+                try:
+                    print("Before socket")
```
Can this line be deleted?  If not, then at least turn it into a `logging.debug()` statement.
```suggestion
```

michaelsaki replied:
@dav3r Yes it can. Sorry it was some debugging stuff that I had left in.

Commit:
Update src/py3.8/Pipfile

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

Commit:
Update src/py3.9/Pipfile

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,231 @@
+
+        for cve_url in cve_urls:
+            attempts = 0
+            retries = 10
```
The `10` here should be turned into a variable (e.g. something like `MAX_CVE_URL_RETRIES`) and moved up top (by `DEFAULT_NVD_URL` and friends) for easier updates.

michaelsaki replied:
@dav3r I addressed this here https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/d184fbd02d1b43d67c1c9cfa71d4f8b3fb08319e

dav3r commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,231 @@
+                        cve_url,
+                    )
+                    attempts += 1
+                    await asyncio.sleep(5)
```
The `5` here should be turned into a variable (e.g. `CVE_URL_RETRY_WAIT_SEC`) and moved up top (by `DEFAULT_NVD_URL` and friends) for easier updates.

michaelsaki replied:
@dav3r I addressed this here: https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/d184fbd02d1b43d67c1c9cfa71d4f8b3fb08319e

dav3r commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,231 @@
 """Simple AWS Lambda handler to verify functionality."""
```
The docstring for this file should be updated from the skeleton:
```suggestion
"""AWS Lambda handler to sync NVD data with a MongoDB."""
```

michaelsaki replied:
@dav3r Addressed this here https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/7169a0a03d4a1758e3fdc9f8a92490175a18c8c3

dav3r replied:
This convo should've been automatically resolved when you committed my suggestion.  Not sure what happened here.

michaelsaki replied:
@dav3r I was thinking the original commenter needed to approve the changes. So I had unresolved it but I will just leave it resolved in the future. Just wanted to make sure I followed processes correctly

dav3r replied:
Your thinking was correct- this is the only exception to our "resolve conversations" rule.  When a suggestion is accepted, GitHub automatically resolves the conversation on behalf of the original commenter since you are committing exactly what the original commenter suggested.

dav3r commented on src/lambda_handler.py:
```diff
@@ -109,19 +253,78 @@ def handler(event, context) -> dict[str, Optional[str]]:
+        motor_client = AsyncIOMotorClient(mongodb_uri)
+
+    try:
+        # This could be optimized to run in parallel using asyncio.gather() but the lambda handler
+        # is currently configured to run in a single-threaded environment so we should come back
+        # to this later when the lambda is configured to run in an asynchronous environment.
```
Is this comment now outdated?

michaelsaki replied:
@dav3r This was a comment that Shane had suggested we leave in so we can come back and get the `handler` function to run async too. See here : https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2#discussion_r1300551086. I could reword it a bit differently if we think it might not exactly make sense to someone reading it in the future

dav3r replied:
I did find it confusing since we are doing some asynchronous activities within this Lambda handler.  Can you take a whack at improving that comment?

jsf9k replied:
My objection is to the fact that we actually have to touch `asyncio.get_event_loop()`.  This is pretty ugly, and I wanted a comment left to remind us to find a better solution.

dav3r replied:
Can we just get an issue in this repo that clearly explains the situation and then link to that issue here?

jsf9k replied:
The `asyncio.get_event_loop()` call is icky because, as stated [here](https://docs.python.org/3.11/library/asyncio-eventloop.html), "Application developers should typically use the high-level asyncio functions, such as [asyncio.run()](https://docs.python.org/3.11/library/asyncio-runner.html#asyncio.run), and should rarely need to reference the loop object or call its methods. This section is intended mostly for authors of lower-level code, libraries, and frameworks, who need finer control over the event loop behavior."

michaelsaki replied:
@jsf9k and @dav3r I addressed this here https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/58fed5f85c6eed73a804749a03e60003d731c799 also I made a new issue https://github.com/cisagov/cyhy-nvdsync-lambda/issues/4

jsf9k replied:
I'm happy now.  @dav3r, you can resolve this conversation as far as I am concerned.

dav3r replied:
Thanks gents!

dav3r reviewed (changes requested):
Please take a look at these additional items I noted.

Commit:
Update src/lambda_handler.py

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

Commit:
Update docstring for the lambda handler description

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k commented:
@michaelsaki - When you do accept the suggested changes, please change the commit message from the default value (e.g., "Update src/lambda_handler.py") to something more specific that actually describes the change or changes being made.

Commit:
Add constant variables

Commit:
Update error handling to catch malformed JSON earlier

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,231 @@
+NVD_FIRST_YEAR = 2002
+DEFAULT_NVD_COLLECTION = "cves"
+MAX_CVE_URL_RETRIES = 10
+CVE_URL_RETRY_WAIT_SEC = 5
```
This group of constants is now large enough that their non-alphabetical order is mildly irritating to my eyes:
```suggestion
CVE_URL_RETRY_WAIT_SEC = 5
DEFAULT_NVD_COLLECTION = "cves"
DEFAULT_NVD_URL = "https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-{year}.json.gz"
MAX_CVE_URL_RETRIES = 10
NVD_FIRST_YEAR = 2002
```

Commit:
Alphabetize constant variables for better readability

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

Commit:
Clarify comment on get_event_loop issue

jsf9k commented on src/lambda_handler.py:
```diff
@@ -109,19 +253,79 @@ def handler(event, context) -> dict[str, Optional[str]]:
+        motor_client = AsyncIOMotorClient(mongodb_uri)
+
+    try:
+        # Using asyncio.get_event_loop() here because the subsequent runs will lead to
+        # a closed event loop error. This needs to be fixed in the future. An issue has been created
+        # for this.
+        # https://github.com/cisagov/cyhy-nvdsync-lambda/issues/4
```
```suggestion
        # TODO:  Using asyncio.get_event_loop() here because otherwise subsequent runs will lead to
        # a closed event loop error. This needs to be fixed in the future. An issue has been created
        # for this:
        # https://github.com/cisagov/cyhy-nvdsync-lambda/issues/4
```

Commit:
Edit comment's grammar and make a TODO

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r reviewed (approved):
LGTM - good work here @michaelsaki 👍 👍

michaelsaki commented:
@jsf9k and @dav3r Thanks for all the awesome feedback and suggestions!

jsf9k reviewed (approved):
Strong work!  Thanks for sticking with us through the review process.

mcdonnnj commented:
Just a reminder that I'm currently working to integrate this into [cisagov/cyhy_amis](https://github.com/cisagov/cyhy_amis) and as part of that test a deployed Lambda artifact. I've run into some minor changes I'd like to see but want to get to a successful deploy before making any change requests.

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -109,19 +253,79 @@ def handler(event, context) -> dict[str, Optional[str]]:
+        return
+
+    # Determine the database where the CVE data will be inserted
+    write_db = get_ssm_parameter(os.environ.get("ssm_db_writedb", "ssm_db_authdb"))
```
This needs to be the value of the `ssm_db_authdb` environment variable.

```suggestion
    write_db = get_ssm_parameter(os.environ.get("ssm_db_writedb", os.environ.get("ssm_db_authdb", "")))
```

michaelsaki replied:
@mcdonnnj  Linting is throwing up errors because it doesn't have a default string value. I was thinking of throwing in a `None` so it won't be an issue. I figured the error checking that happens before this is called would catch any unset environment variables anyways. So it would look like `write_db = get_ssm_parameter(os.environ.get("ssm_db_writedb", os.environ.get("ssm_db_authdb", None)))`

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -109,19 +253,79 @@ def handler(event, context) -> dict[str, Optional[str]]:
+        cve_json_urls = [json_url]
+
+    # Determine if a non-default collection is being used
+    db_collection = os.environ.get("target_collection")
```
Since this is stored in SSM Parameter Store we should name it accordingly for consistency.

```suggestion
    db_collection = os.environ.get("ssm_db_target_collection")
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -109,19 +253,79 @@ def handler(event, context) -> dict[str, Optional[str]]:
+        "ssm_db_user",
+        "ssm_db_pass",
+        "ssm_db_host",
+        "ssm_db_port",
```
Moving these away from SSM since they are not particularly sensitive and wouldn't otherwise be stored in SSM Parameter Store. I'm pretty sure this was sourced from code I wrote so apologies for the change request, but when implementing the Terraform infrastructure to deploy this it felt needlessly convoluted.

```suggestion
        "db_host",
        "db_port",
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,231 @@
+
+
+def build_mongodb_uri(
+    ssm_username: str, ssm_password: str, ssm_host: str, ssm_port: str, ssm_auth_db: str
```
```suggestion
    ssm_username: str, ssm_password: str, db_host: str, db_port: str, ssm_auth_db: str
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -109,19 +253,79 @@ def handler(event, context) -> dict[str, Optional[str]]:
+        "ssm_db_port",
+        "ssm_db_authdb",
+    ]:
+        mongodb_uri_elements.append((var, os.environ.get(var)))
```
This function expects string values so we should ensure that is what it gets.

```suggestion
        mongodb_uri_elements.append((var, os.environ.get(var, "")))
```

mcdonnnj commented on src/lambda_handler.py:
```diff
@@ -1,89 +1,231 @@
+        uri = f"mongodb://{username}:{password}@{host}:{port}/{auth_db}"
+    except ClientError as client_err:
+        logging.error("Unable to create MongoDB URI.")
+        logging.exception(client_err)
```
Adjusting this since two values are no longer stored in SSM Parameter Store.

```suggestion
        # Ensure the database host and port were provided
        if not db_host:
            raise ValueError("The database hostname must be provided.")
        if not db_port:
            raise ValueError("The database port must be provided.")
        # Ensure the username and password are safely encoded
        username = urllib.parse.quote_plus(get_ssm_parameter(ssm_username))
        password = urllib.parse.quote_plus(get_ssm_parameter(ssm_password))
        auth_db = get_ssm_parameter(ssm_auth_db)
        uri = f"mongodb://{username}:{password}@{host}:{port}/{auth_db}"
    except (ClientError, ValueError) as err:
        logging.error("Unable to create MongoDB URI.")
        logging.exception(err)
```

Commit:
Add value of the ssm_db_authdb environment variable

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

Commit:
Rename target_collection to reflect SSM param store variable

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

Commit:
Rename variables to db_host and db_port

"Moving these away from SSM since they are not particularly sensitive and wouldn't otherwise be stored in SSM Parameter Store....when implementing the Terraform infrastructure to deploy this it felt needlessly convoluted."

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

Commit:
Rename more instances of ssm_host and ssm_port

Should have batched all the changes together for these variables from previous commit

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

Commit:
Add parameter for string values being passed in

This function expects string values so we should ensure that is what it gets.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

Commit:
Update host and port to db_host and db_port

Commit:
Add empty default string

Commit:
Add formatting changes from black

michaelsaki commented:
@mcdonnnj Were there any other changes that needed to be made to get this PR merged or issues?

mcdonnnj commented:
> @mcdonnnj Were there any other changes that needed to be made to get this PR merged or issues?

@michaelsaki I'm still working on the Terraform configuration to test this as deployed. I may need to make some additional change requests based on how that goes.
